cmd/util/cmd/find-trie-root/cmd.go

```diff
@@ -0,0 +1,379 @@
+		}
+
+		lg = lg.With().
+			Uint8("operation", uint8(operation)).
+			Logger()
+
+		switch operation {
+		case wal.WALUpdate:
+			rootHash := update.RootHash
+
+			lg.Debug().
+				Str("root-hash", rootHash.String()).
+				Msg("found WALUpdate")
+
+			if rootHash.Equals(expectedHash) {
+				log.Info().Msgf("found expected trie root hash %v", rootHash)
+				return reader.Segment(), reader.Offset(), nil
+			}
+		default:
+		}
+
+		err = reader.Err()
+		if err != nil {
+			return 0, 0, fmt.Errorf("cannot read LedgerWAL: %w", err)
+		}
+	}
+
+	return 0, 0, fmt.Errorf("finish reading all segment files from %d to %d, but not found", from, to)
+}
+
+// findRootHashAndCreateTrimmed finds the root hash in the segment file from the given dir folder
+// and creates a new segment file with the expected root hash as the last record in a temporary folder.
+// it return the path to the new segment file.
+func findRootHashAndCreateTrimmed(dir string, segment int, expectedRoot ledger.RootHash) (string, error) {
+	tmpFolder, err := os.MkdirTemp("", "flow-last-segment-file")
+	if err != nil {
+		return "", fmt.Errorf("cannot create temporary folder: %w", err)
+	}
+
+	// the new segment file will be created in the temporary folder
+	// and it's always 00000000
+	newSegmentFile := prometheusWAL.SegmentName(tmpFolder, 0)
+
+	log.Info().Msgf("writing new segment file to %v", newSegmentFile)
+
+	writer, err := prometheusWAL.NewSize(log.Logger, nil, tmpFolder, wal.SegmentSize, false)
+	if err != nil {
+		return "", fmt.Errorf("cannot create writer WAL: %w", err)
+	}
+
+	defer writer.Close()
+
+	sr, err := prometheusWAL.NewSegmentsRangeReader(log.Logger, prometheusWAL.SegmentRange{
+		Dir:   dir,
+		First: segment,
+		Last:  segment,
+	})
+	if err != nil {
+		return "", fmt.Errorf("cannot create WAL segments reader: %w", err)
+	}
+
+	defer sr.Close()
+
+	reader := prometheusWAL.NewReader(sr)
+
+	for reader.Next() {
+		record := reader.Record()
+		operation, _, update, err := wal.Decode(record)
+		if err != nil {
+			return "", fmt.Errorf("cannot decode LedgerWAL record: %w", err)
+		}
+
+		switch operation {
+		case wal.WALUpdate:
+
+			bytes := wal.EncodeUpdate(update)
+			_, err = writer.Log(bytes)
+			if err != nil {
+				return "", fmt.Errorf("cannot write LedgerWAL record: %w", err)
+			}
+
+			rootHash := update.RootHash
+
+			if rootHash.Equals(expectedRoot) {
+				log.Info().Msgf("found expected trie root hash %v, finish writing", rootHash)
+				return newSegmentFile, nil
+			}
+		default:
+		}
+
+		err = reader.Err()
+		if err != nil {
+			return "", fmt.Errorf("cannot read LedgerWAL: %w", err)
+		}
+	}
+
+	return "", fmt.Errorf("finish reading all segment files from %d to %d, but not found", segment, segment)
+}
+
+func checkFolderNotExistOrEmpty(folderPath string) (bool, error) {
+	// Check if the folder exists
+	info, err := os.Stat(folderPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return true, nil
+		}
+		return false, nil
```

Comment:
Should we return error here?

```suggestion
		return false, err
```